src/Raven.Server/Smuggler/Migration/Migrator.cs

```diff
@@ -0,0 +1,354 @@
+    public class Migrator
+    {
+        private const string MigrationStateKeyBase = "Raven/Migration/Status";
+
+        private readonly HttpClient _client;
+        private readonly string _serverUrl;
+        private readonly string _migrationStateKey;
+        private readonly ServerStore _serverStore;
+        private readonly Operations _operations;
+        private readonly CancellationToken _cancellationToken;
+
+        public Migrator(
+            MigrationConfigurationBase configuration, 
+            ServerStore serverStore,
+            Operations operations,
+            CancellationToken cancellationToken)
+        {
+            var hasCredentials =
+                string.IsNullOrWhiteSpace(configuration.UserName) == false &&
+                string.IsNullOrWhiteSpace(configuration.Password) == false;
+
+            var httpClientHandler = new HttpClientHandler
+            {
+                UseDefaultCredentials = hasCredentials == false,
+            };
+            if (hasCredentials)
+                httpClientHandler.Credentials = new NetworkCredential(
+                    configuration.UserName,
+                    configuration.Password,
+                    configuration.Domain);
+
+            _client = new HttpClient(httpClientHandler);
+
+            _serverUrl = configuration.ServerUrl.TrimEnd('/');
+            _migrationStateKey = $"{MigrationStateKeyBase}/{_serverUrl}";
+            _serverStore = serverStore;
+            _operations = operations;
+            _cancellationToken = cancellationToken;
+        }
+
+        public async Task MigrateDatabases(List<string> databasesToMigrate)
+        {
+            if (databasesToMigrate == null || databasesToMigrate.Count == 0)
+            {
+                // migrate all databases
+                databasesToMigrate = await GetDatabasesToMigrate();
+            }
+
+            if (databasesToMigrate.Count == 0)
+                throw new InvalidOperationException("Found no databases to migrate");
+
+            _serverStore.EnsureNotPassive();
+
+            foreach (var databaseNameToMigrate in databasesToMigrate)
+            {
+                await CreateDatabaseIfNeeded(databaseNameToMigrate);
+                var database = await GetDatabase(databaseNameToMigrate);
+                if (database == null)
+                {
+                    // database doesn't exist
+                    continue;
+                }
+
+                StartMigrateSingleDatabase(databaseNameToMigrate, database);
+            }
+        }
+
+        public long StartMigrateSingleDatabase(string sourceDatabaseName, DocumentDatabase database)
+        {
+            var operationId = _operations.GetNextOperationId();
+            var cancelToken = new OperationCancelToken(_cancellationToken);
+            var result = new SmugglerResult();
+
+            _operations.AddOperation(null,
+                $"Migrating database: {sourceDatabaseName} from url: {_serverUrl}",
```

Comment:
Use present simple here. Please have a look at import/export database operations. After migration is completed we have  Database import
September 18, 2017 10:47 AM
Migrating database: Northwind from url: http://live-test.ravendb.net

Reply:
done